### PR TITLE
Suppresses auto-focus under certain conditions

### LINKED ISSUE #8251: Prevent window focus stealing for auto focus scenarios
## Steps To Reproduce
Provide a detailed list of steps that reproduce the issue.

1. Create a react-native-windows app on XAML islands that creates multiple windows on a single UI thread
2. Trigger auto focus for any of the following scenarios:
  a. FlyoutViewManager FlyoutBase instance Closing callback
  b. Call `View.focus()` imperatively from React Native JS
  c. Render a TextInput on a non-foregrounded window with autoFocus={true}

## Expected Results

It's not always clear what the expected results should be. For example, scenario 2a above can be incredibly annoying when the flyout closes when you switch windows (the focus immediately returns to the window you were trying to focus away from).

Scenario 2b and 2c are a bit less clear. Perhaps you want the window to focus when you imperatively focus a control in 2b. I doubt we'd ever want autoFocus to take focus the wrong window for TextInput autoFocus in 2c.

However, I think the developer needs to be afforded the opportunity to decide if they want the window to get focus. For example, if you want to imperatively focus a control on a non-foreground window, you should likely focus the window first, then the control. If we leave things the way they are, then the developer is always forced to focus the window simultaneously with the control.

### PR DESCRIPTION
**Pull request by rozele**
In multi-window XAML Islands apps that share a single UI thread for each window, focusing a view on a blurred window causes that window to get focus. Generally speaking, focusing a window should be explicit, and it's likely very rare that you would want a view getting focus to cause the entire window to get focus.

This manifests as problematic in two ways:
1. If you have a flyout shown and the flyout gets dismissed because you switch to another window, the parent window for the flyout would steal focus again when react-native-windows attempts to focus the target element.
2. If you re-render a view with a TextInput with `autoFocus={true}` and the parent window for that TextInput is blurred, the window will steal focus.

This change forces the user to first focus the window and update the active XamlRoot via the `XamlUIService::SetXamlRoot` API before a view in the window can get focus. In other words, the user has to explicitly focus the window before a programmatic view focus behavior can cause the window to get focused.

Fixes #8251 

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/8393)

### CODE REVIEW

**Comment by asklar** [comment]
I'm not sure this fix is the right behavior. The reason that the window steals focus is that it's running on the same thread. if the window were to be running on a different thread, then the window would not get focus (unless you make an explicit choice to allow another thread to get focus), and instead its taskbar icon would flash indicating action is required. So I believe what you are seeing is an artifact of running several [xaml island] windows on the same thread when all of the windows are top level. You might want to scope down the change behind a quirk, as other apps might not want to be artificially limited in not being able to switch focus to another window on the same thread.

**Comment by acoates-ms** [comment]
The all top level windows should be on different threads is a UWP thing.  In win32 its fairly common to have multiple top level windows that are all on the same thread.  -- Office included.  This does seem like something that XAML should think about with winUI going forward.  I dont think the thread check is going to be sufficient.  -- But I also dont know how XAML would be able to differentiate between two islands in the same window, and two seperate top windows.

Doing so will be a requirement for Office moving to XAML islands too though.   I agree with @asklar that a quirk in RNW for now makes sense. -- But we should open an issue in XAML to think about this more for the future.

**Comment by rozele** [comment]
@asklar @acoates-ms added a quirk setting. I'll also file a bug on WinUI.

**Comment by NickGerleman** [comment]
@acoates-ms ping to take another look

File: vnext/Microsoft.ReactNative/Views/FlyoutViewManager.cpp [hunk 1]
```diff
@@ -172,12 +172,14 @@ void FlyoutShadowNode::createView(const winrt::Microsoft::ReactNative::JSValueOb
...
     if (!m_updating) {
-      if (m_targetElement != nullptr) {
+      if (m_targetTag > 0) {
         // When the flyout closes, attempt to move focus to
         // its anchor element to prevent cases where focus can land on
         // an outer flyout content and therefore trigger a unexpected flyout
         // dismissal
-        xaml::Input::FocusManager::TryFocusAsync(m_targetElement, winrt::FocusState::Programmatic);
+        if (auto uiManager = GetNativeUIManager(GetViewManager()->GetReactContext()).lock()) {
+          uiManager->focus(m_targetTag);
```
[the remark this answers is not in the document]
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/Views/FlyoutViewManager.cpp:181]: The basic difference is that using Keyboard will usually cause the focus rect to show.  Whereas if you use programmatic, then XAML will look at what kind of focus was last used (pointer vs keyboard), and only show the focus rect if the last focus was keyboard.  The general idea is to not clog up the UI with focus rects when the user is a mouse/touch user.  -- Instead only showing it for keyboard users.

**Comment by acoates-ms** [comment]
@asklar is there a way to tell XAML to only update its internal focus without causing it to activate the window?

This PR seem to completely skip the focus call if the window containing the control doesn't currently have focus.  But I don't think that's the behavior we'd want.  What we really want is to tell XAML that it should have that control focused when that window is activated.  But we dont want to explicitly activate it now.  -- This seems some potentially some fallout from moving XAML out of UWP.  I suspect we may need to add some more fine grained focus methods to WinUI3?

**Comment by rozele** [comment]
@acoates-ms - indeed, this does create a "bug" in that the focus state is not correct on the inactive window. I still think this is better than forcing a window to focus.

**Comment by AgneLukoseviciute** [comment]
I'm going through trying to get our list of open PRs down a bit, this one's been open for a while. Now that it's behind a quirk setting, seems like we might be ok to merge? @asklar @acoates-ms, any remaining concerns?

**Review by AgneLukoseviciute — APPROVED** [review]

**Review by acoates-ms — APPROVED** [review]

**Pull request merged**